ksmt-core/src/main/kotlin/org/ksmt/cache/weak/WeakCacheKey.kt

```diff
@@ -0,0 +1,45 @@
+package org.ksmt.cache.weak
+
+import org.ksmt.cache.KInternedObject
+import java.lang.ref.ReferenceQueue
+import java.lang.ref.WeakReference
+
+internal interface WeakCacheKey<K : KInternedObject> {
+    fun get(): K?
+}
+
+internal class WeakCacheKeyRef<K : KInternedObject>(
+    key: K?, queue: ReferenceQueue<K>?,
+    private val hashCode: Int
+) : WeakReference<K>(key, queue), WeakCacheKey<K> {
+
+    override fun equals(other: Any?): Boolean = when {
+        other === this -> true
+        other is WeakCacheKey<*> -> get() === other.get()
+        else -> false
+    }
+
+    override fun hashCode(): Int = hashCode
```

Comment:
Why do we have guarantees about the connection between equals and `hashCode` provided in the method?